### PR TITLE
chore(integer): brings the CPU and GPU tests into line.

### PR DESCRIPTION
**Pull request by pdroalves**
- also implements Debug, PartialEq to CompressedCiphertextList

closes: _please link all relevant issues_

### Check-list:

* [ ] Tests for the changes have been added (for bug fixes / features)
* [ ] Docs have been added / updated (for bug fixes / features)
* [ ] Relevant issues are marked as resolved/closed, related issues are linked in the description
* [ ] Check for breaking changes (including serialization changes) and add them to commit message following the conventional commit [specification][conventional-breaking]

[conventional-breaking]: https://www.conventionalcommits.org/en/v1.0.0/#commit-message-with-description-and-breaking-change-footer

### CODE REVIEW

File: tfhe/src/integer/gpu/ciphertext/compressed_ciphertext_list.rs [hunk 1]
```diff
@@ -519,7 +519,7 @@ mod tests {
     use rand::Rng;
 
     const NB_TESTS: usize = 10;
-    const NB_OPERATOR_TESTS: usize = 10;
+    const NB_OPERATOR_TESTS: usize = 1;
```
Comment (agnesLeroy) [inline tfhe/src/integer/gpu/ciphertext/compressed_ciphertext_list.rs:522]: I don't think we want to reduce this to 1.

File: tfhe/src/core_crypto/entities/packed_integers.rs [hunk 2]
```diff
@@ -4,7 +4,7 @@ use crate::conformance::ParameterSetConformant;
 use crate::core_crypto::backward_compatibility::entities::packed_integers::PackedIntegersVersions;
 use crate::core_crypto::prelude::*;
 
-#[derive(Clone, serde::Serialize, serde::Deserialize, Versionize)]
+#[derive(Clone, Debug, Eq, PartialEq, serde::Serialize, serde::Deserialize, Versionize)]
```
Comment (agnesLeroy) [inline tfhe/src/core_crypto/entities/packed_integers.rs:7]: Why do you need to derive Debug, Eq, PartialEq?

File: tfhe/src/integer/ciphertext/compressed_ciphertext_list.rs [hunk 3]
```diff
@@ -195,4 +197,173 @@ mod tests {
 
         assert!(cks.decrypt_bool(&decompressed3));
     }
+
+    const NB_TESTS: usize = 10;
+    const NB_OPERATOR_TESTS: usize = 1;
+    #[test]
+    fn test_ciphertext_compression() {
```
Comment (agnesLeroy) [inline tfhe/src/integer/ciphertext/compressed_ciphertext_list.rs:204]: Don't you want to remove the existing test to keep only this one?

File: tfhe/src/integer/ciphertext/compressed_ciphertext_list.rs [hunk 4]
```diff
@@ -195,4 +197,173 @@ mod tests {
...
+        let mut rng = rand::thread_rng();
+
+        let message_modulus: u128 = cks.parameters().message_modulus().0 as u128;
+
+        let num_blocks = 32;
+        for _ in 0..NB_TESTS {
+            // Unsigned
+            let modulus = message_modulus.pow(num_blocks as u32);
+            for _ in 0..NB_OPERATOR_TESTS {
+                let nb_messages = 1 + (rng.gen::<u64>() % 6);
```
Comment (agnesLeroy) [inline tfhe/src/integer/ciphertext/compressed_ciphertext_list.rs:222]: Why do we only test up to 6 nb_messages? Why not more?

File: tfhe/src/integer/ciphertext/compressed_ciphertext_list.rs [hunk 5]
```diff
@@ -195,4 +197,173 @@ mod tests {
...
+                let messages = (0..nb_messages)
+                    .map(|_| rng.gen::<u128>() % modulus)
```
Comment (agnesLeroy) [inline tfhe/src/integer/ciphertext/compressed_ciphertext_list.rs:224]: Here you could generate a 64 bit integer since there are 32 blocks.

**Review by agnesLeroy — COMMENTED** [review]
Thanks a lot @pdroalves! It looks good to me overall, I just have some small comments/questions. :slightly_smiling_face:

File: tfhe/src/integer/gpu/ciphertext/compressed_ciphertext_list.rs (hunk 1 above)
Comment (agnesLeroy) [reply tfhe/src/integer/gpu/ciphertext/compressed_ciphertext_list.rs:522]: Btw shouldn't we increase the number of tests and operator tests for CPU & GPU?

**Comment by IceTDrinker** [comment]
follow up issue was here https://github.com/zama-ai/tfhe-rs-internal/issues/618

File: tfhe/src/integer/ciphertext/compressed_ciphertext_list.rs (hunk 3 above)
Comment (pdroalves) [reply tfhe/src/integer/ciphertext/compressed_ciphertext_list.rs:204]: Indeed.

File: tfhe/src/integer/ciphertext/compressed_ciphertext_list.rs (hunk 4 above)
Comment (pdroalves) [reply tfhe/src/integer/ciphertext/compressed_ciphertext_list.rs:222]: I'm setting it up to `2 * compressed_polynomial_size`.

File: tfhe/src/integer/ciphertext/compressed_ciphertext_list.rs [hunk 6]
```diff
@@ -195,4 +197,173 @@ mod tests {
...
+                    .collect::<Vec<_>>();
+
+                let cts = messages
+                    .iter()
+                    .map(|message| cks.encrypt_radix(*message, num_blocks))
+                    .collect_vec();
```
Comment (agnesLeroy) [inline tfhe/src/integer/ciphertext/compressed_ciphertext_list.rs:230]: Before compression/decompression we should perform a `bitand` between each message and itself to make sure the noise is the nominal one after a PBS, and not just the encryption noise.

File: tfhe/src/integer/ciphertext/compressed_ciphertext_list.rs (hunk 5 above)
Comment (IceTDrinker) [reply tfhe/src/integer/ciphertext/compressed_ciphertext_list.rs:224]: the u128 is fine in the end I did not realize we have generic encrypt/decrypt

Comment (agnesLeroy) [reply tfhe/src/integer/ciphertext/compressed_ciphertext_list.rs:224]: Ah ok sorry

Comment (pdroalves) [reply tfhe/src/integer/ciphertext/compressed_ciphertext_list.rs:224]: When I do that I get an 

```thread 'integer::ciphertext::compressed_ciphertext_list::tests::test_ciphertext_compression' panicked at tfhe/src/integer/ciphertext/compressed_ciphertext_list.rs:188:30:
attempt to calculate the remainder with a divisor of zero
```

 error

File: tfhe/src/core_crypto/entities/packed_integers.rs (hunk 2 above)
Comment (pdroalves) [reply tfhe/src/core_crypto/entities/packed_integers.rs:7]: Maybe we don't need Debug. I added Eq and PartialEq so we can compare `CompressedCiphertextList` and make sure GPU and CPU's compression are identical. I just pushed a modification to `to_cuda_compressed_ciphertext_list` doctest doing that.

File: tfhe/src/integer/ciphertext/compressed_ciphertext_list.rs (hunk 6 above)
Comment (pdroalves) [reply tfhe/src/integer/ciphertext/compressed_ciphertext_list.rs:196]: I understand this should be at the test and not in compression/decompression algorithms, right?

File: tfhe/src/core_crypto/entities/packed_integers.rs (hunk 2 above)
Comment (IceTDrinker) [reply tfhe/src/core_crypto/entities/packed_integers.rs:7]: we can't guarantee that if we have PBSes applied right ? as we may have some slightly different noise depending on fft impls ?

File: tfhe/src/integer/ciphertext/compressed_ciphertext_list.rs (hunk 6 above)
Comment (IceTDrinker) [reply tfhe/src/integer/ciphertext/compressed_ciphertext_list.rs:196]: yes it should be in the test to generate integers with post computation levels of noise

so encrypt -> bitand with self -> compress -> decompress -> decrypt -> check decrypted value

Comment (pdroalves) [reply tfhe/src/integer/ciphertext/compressed_ciphertext_list.rs:196]: If I add this line to CPU's these the test will fail every time:

```
                for ct in cts {
                    let and_ct = radix_sks.bitand_parallelized(&ct, &ct);
                    builder.push(and_ct);
                }
```

Comment (pdroalves) [reply tfhe/src/integer/ciphertext/compressed_ciphertext_list.rs:196]: We should probably wait for the fix before doing this, or completely remove this test from the CI in the meantime.

File: tfhe/src/core_crypto/entities/packed_integers.rs (hunk 2 above)
Comment (pdroalves) [reply tfhe/src/core_crypto/entities/packed_integers.rs:7]: Current implementation only has a PBS for decompression. Compression's result should be exactly the same.

File: tfhe/src/integer/ciphertext/compressed_ciphertext_list.rs (hunk 6 above)
Comment (IceTDrinker) [reply tfhe/src/integer/ciphertext/compressed_ciphertext_list.rs:196]: The goal here is: pessimize the CI and have the proper setting, wait for the fix and fix here

Comment (pdroalves) [reply tfhe/src/integer/ciphertext/compressed_ciphertext_list.rs:200]: I see. In this case I'm adding the bitand operation before compress.

File: tfhe/src/integer/ciphertext/compressed_ciphertext_list.rs [hunk 7]
```diff
@@ -153,46 +153,188 @@ impl CompressedCiphertextList {
...
+            for _ in 0..NB_OPERATOR_TESTS {
+                let nb_messages = 1 + (rng.gen::<u64>() % MAX_NB_MESSAGES as u64);
+                let messages = (0..nb_messages)
+                    .map(|_| rng.gen::<i128>() % modulus)
+                    .collect::<Vec<_>>();
+
+                let cts = messages
+                    .iter()
+                    .map(|message| cks.encrypt_signed_radix(*message, NUM_BLOCKS))
+                    .collect_vec();
```
Comment (agnesLeroy) [inline tfhe/src/integer/ciphertext/compressed_ciphertext_list.rs:229]: We need to compute the bitand here as well

File: tfhe/src/integer/ciphertext/compressed_ciphertext_list.rs [hunk 8]
```diff
@@ -153,46 +153,188 @@ impl CompressedCiphertextList {
...
+                let nb_messages = 1 + (rng.gen::<u64>() % MAX_NB_MESSAGES as u64);
+                let messages = (0..nb_messages)
+                    .map(|_| rng.gen::<i64>() % 2 != 0)
+                    .collect::<Vec<_>>();
+
+                let cts = messages
+                    .iter()
+                    .map(|message| cks.encrypt_bool(*message))
+                    .collect_vec();
+
```
Comment (agnesLeroy) [inline tfhe/src/integer/ciphertext/compressed_ciphertext_list.rs:257]: And here

File: tfhe/src/integer/ciphertext/compressed_ciphertext_list.rs [hunk 9]
```diff
@@ -153,46 +153,188 @@ impl CompressedCiphertextList {
...
+                            let message = rng.gen::<u128>() % modulus;
+                            let ct = cks.encrypt_radix(message, NUM_BLOCKS);
+                            builder.push(ct);
+                            messages.push(MessageType::Unsigned(message));
+                        }
+                        1 => {
+                            // Signed
+                            let modulus = message_modulus.pow((NUM_BLOCKS - 1) as u32) as i128;
+                            let message = rng.gen::<i128>() % modulus;
+                            let ct = cks.encrypt_signed_radix(message, NUM_BLOCKS);
```
Comment (agnesLeroy) [inline tfhe/src/integer/ciphertext/compressed_ciphertext_list.rs:299]: Here as well

File: tfhe/src/integer/ciphertext/compressed_ciphertext_list.rs [hunk 10]
```diff
@@ -153,46 +153,188 @@ impl CompressedCiphertextList {
...
+                            builder.push(ct);
+                            messages.push(MessageType::Signed(message));
+                        }
+                        _ => {
+                            // Boolean
+                            let message = rng.gen::<i64>() % 2 != 0;
+                            let ct = cks.encrypt_bool(message);
```
Comment (agnesLeroy) [inline tfhe/src/integer/ciphertext/compressed_ciphertext_list.rs:306]: And here

File: tfhe/src/integer/ciphertext/compressed_ciphertext_list.rs [hunk 11]
```diff
@@ -153,46 +153,188 @@ impl CompressedCiphertextList {
...
     use super::*;
-    use crate::integer::ClientKey;
+    use crate::integer::{gen_keys_radix, ClientKey};
     use crate::shortint::parameters::list_compression::COMP_PARAM_MESSAGE_2_CARRY_2_KS_PBS_TUNIFORM_2M64;
     use crate::shortint::parameters::PARAM_MESSAGE_2_CARRY_2_KS_PBS_TUNIFORM_2M64;
+    use itertools::Itertools;
+    use rand::Rng;
 
+    const NB_TESTS: usize = 10;
+    const NB_OPERATOR_TESTS: usize = 10;
```
Comment (agnesLeroy) [inline tfhe/src/integer/ciphertext/compressed_ciphertext_list.rs:163]: The nb_tests and nb_operator_tests should be set to 100

File: tfhe/src/integer/ciphertext/compressed_ciphertext_list.rs [hunk 12]
```diff
@@ -153,46 +153,188 @@ impl CompressedCiphertextList {
...
     #[test]
-    fn test_heterogeneous_ciphertext_compression_ci_run_filter() {
+    fn test_ciphertext_compression() {
+        const NUM_BLOCKS: usize = 32;
+
         let cks = ClientKey::new(PARAM_MESSAGE_2_CARRY_2_KS_PBS_TUNIFORM_2M64);
 
+        let (_, radix_sks) =
+            gen_keys_radix(PARAM_MESSAGE_2_CARRY_2_KS_PBS_TUNIFORM_2M64, NUM_BLOCKS);
```
Comment (IceTDrinker) [inline tfhe/src/integer/ciphertext/compressed_ciphertext_list.rs:171]: this is wrong you are generating a server key taht does not correspond to the encryption client key

File: tfhe/src/integer/ciphertext/compressed_ciphertext_list.rs (hunk 7 above)
Comment (pdroalves) [reply tfhe/src/integer/ciphertext/compressed_ciphertext_list.rs:229]: Of course. I forgot about the other types.

File: tfhe/src/integer/ciphertext/compressed_ciphertext_list.rs [hunk 13]
```diff
@@ -153,46 +153,188 @@ impl CompressedCiphertextList {
...
+                    .map(|message| cks.encrypt_bool(*message))
+                    .collect_vec();
+
+                let mut builder = CompressedCiphertextListBuilder::new();
+
+                for ct in cts {
+                    builder.push(ct);
+                }
+
+                let cuda_compressed = builder.build(&compression_key);
```
[the remark this answers is not in the document]
Comment (pdroalves) [reply tfhe/src/integer/ciphertext/compressed_ciphertext_list.rs]: Ops. Removed.

File: tfhe/src/integer/gpu/ciphertext/compressed_ciphertext_list.rs [hunk 14]
```diff
@@ -287,46 +287,49 @@ impl CompressedCiphertextList {
...
+    ///         cuda_compressed.get(1, &cuda_decompression_key, &streams).unwrap().unwrap();
+    ///     let decompressed2 = d_decompressed2.to_signed_radix_ciphertext(&streams);
+    ///     let decrypted: i32 = radix_cks.decrypt_signed(&decompressed2);
+    ///     assert_eq!(decrypted, -2);
+    ///
+    ///     let d_decompressed3: CudaBooleanBlock =
+    ///         cuda_compressed.get(2, &cuda_decompression_key, &streams).unwrap().unwrap();
+    ///     let decompressed3 = d_decompressed3.to_boolean_block(&streams);
+    ///     let decrypted = radix_cks.decrypt_bool(&decompressed3);
+    ///     assert!(decrypted);
```
Comment (IceTDrinker) [inline tfhe/src/integer/gpu/ciphertext/compressed_ciphertext_list.rs:332]: why the indentation here ?

File: Makefile [hunk 15]
```diff
@@ -552,6 +552,13 @@ test_integer_gpu: install_rs_build_toolchain
...
 		--features=$(TARGET_ARCH_FEATURE),integer,gpu -p $(TFHE_SPEC) -- integer::gpu::server_key::
 
+.PHONY: test_integer_compression
+test_integer_compression: install_rs_build_toolchain
+	RUSTFLAGS="$(RUSTFLAGS)" cargo $(CARGO_RS_BUILD_TOOLCHAIN) test --profile $(CARGO_PROFILE) \
+		--features=$(TARGET_ARCH_FEATURE),integer -p $(TFHE_SPEC) -- integer::ciphertext::compressed_ciphertext_list::tests::
+	RUSTFLAGS="$(RUSTFLAGS)" cargo $(CARGO_RS_BUILD_TOOLCHAIN) test --doc --profile $(CARGO_PROFILE) \
+		--features=$(TARGET_ARCH_FEATURE),integer -p $(TFHE_SPEC) -- integer::ciphertext::compress
+
 .PHONY: test_integer_compression_gpu
```
Comment (IceTDrinker) [inline Makefile:562]: this is "just" a convenience target right ? the tests run in normal workflows ?

Comment (pdroalves) [reply Makefile:562]: This is a convenience target. Also, the GPU compression test runs within the CUDA fast test workflow. I'm not sure the CPU test has ever been on any workflow. I would say not.

File: tfhe/src/integer/gpu/ciphertext/compressed_ciphertext_list.rs (hunk 14 above)
Comment (pdroalves) [reply tfhe/src/integer/gpu/ciphertext/compressed_ciphertext_list.rs:332]: For some reason `make fmt_gpu` will not correctly indent this part. I tried to fix indentation manually.

File: Makefile (hunk 15 above)
Comment (IceTDrinker) [reply Makefile:562]: it should be in the Unsigned integer tests, the cargo nextest run should have it, I'll recheck the logs

Comment (pdroalves) [reply Makefile:562]: I've found it in the unsigned integer. I think it's alright.

**Comment by IceTDrinker** [comment]
we should have a fix for this, looks like an erratum in parameters, we'll let you know once we decide on the way forward

File: tfhe/src/integer/gpu/ciphertext/compressed_ciphertext_list.rs (hunk 14 above)
Comment (IceTDrinker) [reply tfhe/src/integer/gpu/ciphertext/compressed_ciphertext_list.rs]: still broken it seems, I'll try to see if I can fix it

Comment (agnesLeroy) [reply tfhe/src/integer/gpu/ciphertext/compressed_ciphertext_list.rs]: I fixed it manually, but this file doesn't seem to be formatted at all when doing make fmt_gpu, this is strange.

Comment (IceTDrinker) [reply tfhe/src/integer/gpu/ciphertext/compressed_ciphertext_list.rs]: does it have a very long line ? rust fmt sometimes has issues with taht

File: tfhe/src/integer/ciphertext/compressed_ciphertext_list.rs [hunk 16]
```diff
@@ -153,46 +153,193 @@ impl CompressedCiphertextList {
...
-
-        let decrypted2: i32 = cks.decrypt_signed_radix(&decompressed2);
-
-        assert_eq!(decrypted2, -2);
-
-        let decompressed3 = compressed.get(2, &decompression_key).unwrap().unwrap();
-
-        assert!(cks.decrypt_bool(&decompressed3));
+        const MAX_NB_MESSAGES: usize = 2 * COMP_PARAM_MESSAGE_2_CARRY_2_KS_PBS_TUNIFORM_2M64
+            .packing_ks_polynomial_size
```
Comment (IceTDrinker) [inline tfhe/src/integer/ciphertext/compressed_ciphertext_list.rs:180]: lwe_per_glwe

File: tfhe/src/integer/ciphertext/compressed_ciphertext_list.rs [hunk 17]
```diff
@@ -153,46 +153,193 @@ impl CompressedCiphertextList {
...
+
+        let mut rng = rand::thread_rng();
+
+        let message_modulus: u128 = cks.parameters().message_modulus().0 as u128;
+
+        for _ in 0..NB_TESTS {
+            // Unsigned
+            let modulus = message_modulus.pow(NUM_BLOCKS as u32);
+            for _ in 0..NB_OPERATOR_TESTS {
+                let nb_messages = 1 + (rng.gen::<u64>() % MAX_NB_MESSAGES as u64);
```
Comment (IceTDrinker) [inline tfhe/src/integer/ciphertext/compressed_ciphertext_list.rs:192]: use gen_range to make it easier to read

**Review by agnesLeroy — APPROVED** [review]

**Review by IceTDrinker — APPROVED** [review]

**Pull request merged**